fl4health/losses/mkmmd_loss.py

```diff
@@ -77,156 +80,218 @@ def construct_quadruples(self, X: torch.Tensor, Y: torch.Tensor) -> torch.Tensor
-        hat_d_per_kernel = self.compute_hat_d_per_kernel(all_h_u_per_vi)
-        # Take the dot product between the indivudal kernel hat_d values to scale by the basis coefficients
+            all_h_u_per_sample = self.compute_all_h_u_all_samples(X, Y)
+            hat_d_per_kernel = self.compute_hat_d_per_kernel(all_h_u_per_sample)
+        # Take the dot product between the individual kernel hat_d values to scale by the basis coefficients
         return torch.mm(beta.t(), hat_d_per_kernel)[0][0]
 
-    def create_h_u_delta_w_i(self, all_h_u_per_v_i: torch.Tensor) -> torch.Tensor:
-        # Implementing the new version of the h_u_delta_w_i calculation based on David's Notes
-        all_E_h_u_per_v_i = torch.mean(all_h_u_per_v_i, dim=(1, 2), keepdim=True)
+    def form_h_u_delta_w_i(self, all_h_u_per_v_i: torch.Tensor) -> torch.Tensor:
+        # all_h_u_per_v_i has dimension number kernels x number of v_i quadruples
+        n_kernels, n_v_i_s = all_h_u_per_v_i.shape
+        # if the number of v_is is not divisible by two just drop the final ones
+        if n_v_i_s % 2 == 1:
+            all_h_u_per_v_i = all_h_u_per_v_i[:, :-1]
         # pairing the h_u(v_i) values as h_u(v_{2i-1}), h_u(v_{2i}), think w_is from the paper
-        return all_h_u_per_v_i - all_E_h_u_per_v_i
+        h_u_v_i_pairs = all_h_u_per_v_i.reshape(n_kernels, n_v_i_s // 2, 2)
+        return h_u_v_i_pairs[:, :, 0] - h_u_v_i_pairs[:, :, 1]
 
-    def compute_hat_Q_k(self, all_h_u_per_v_i: torch.Tensor) -> torch.Tensor:
+    def compute_hat_Q_k_linear(self, all_h_u_per_v_i: torch.Tensor) -> torch.Tensor:
         # all_h_u_per_v_i has dimension number kernels x number of v_i quadruples
-        h_u_delta_w_i = self.create_h_u_delta_w_i(all_h_u_per_v_i)
+        h_u_delta_w_i = self.form_h_u_delta_w_i(all_h_u_per_v_i)
 
         Q_k_matrix: torch.Tensor = torch.zeros((self.kernel_num, self.kernel_num)).to(self.device)
         len_w_is = h_u_delta_w_i.shape[1]
-        # For each basis function we're adding in the value of h_{j, \Delta}(w_i)*h_{k, \Delta}(w_i) from the
-        # construction above to the proper entry in Q. Note that Q is symmetric. So we can construct the symmetric
-        # entries at the same time.
-        for j in range(self.kernel_num):
-            for k in range(j + 1):
-                Q_k_matrix[j][k] += torch.sum(h_u_delta_w_i[j] * h_u_delta_w_i[k])
-                if j != k:
-                    Q_k_matrix[k][j] += torch.sum(h_u_delta_w_i[j] * h_u_delta_w_i[k])
+        for i in range(len_w_is):
+            # For each basis function we're adding in the value of h_{j, \Delta}(w_i)*h_{k, \Delta}(w_i) from the
+            # construction above to the proper entry in Q. Note that Q is symmetric. So we can construct the symmetric
+            # entries at the same time.
+            for j in range(self.kernel_num):
+                for k in range(j + 1):
+                    Q_k_matrix[j][k] += h_u_delta_w_i[j][i] * h_u_delta_w_i[k][i]
```

Comment:
Why don't we use torch.sum for second dim instead of having the first for loop?

Reply:
I don't know if I can exactly picture the operation you're suggesting, but if its mathematically equivalent, I'm all for not doing three nested loops 😂. Maybe you can write up the operation you're envisioning and I can test out its equivalence?

Reply:
basically this old version I had:
```
     for j in range(self.kernel_num):
            for k in range(j + 1):
                Q_k_matrix[j][k] += torch.sum(h_u_delta_w_i[j] * h_u_delta_w_i[k])
                if j != k:
                    Q_k_matrix[k][j] += torch.sum(h_u_delta_w_i[j] * h_u_delta_w_i[k])
```

Reply:
Ah right. Sorry, I didn't mean to blow that away. I'll mock up a quick test just to make sure they are numerically equivalent and then I'll add it back in.

Reply:
Tested the equivalence and we're good to go. Loops have been updated.